01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/service/PodplayMediaCallback.kt

```diff
@@ -0,0 +1,308 @@
+  private var newMedia: Boolean = false
+  private var mediaExtras: Bundle? = null
+  private var focusRequest: AudioFocusRequest? = null
+  private var mediaNeedsPrepare: Boolean = false
+
+  override fun onCommand(command: String?, extras: Bundle?,
+                         cb: ResultReceiver?) {
+    super.onCommand(command, extras, cb)
+    when (command) {
+      CMD_CHANGESPEED -> extras?.let { changeSpeed(it) }
+    }
+  }
+
+  override fun onPlayFromUri(uri: Uri?, extras: Bundle?) {
+    super.onPlayFromUri(uri, extras)
+
+    if (mediaUri == uri) {
+      newMedia = false
+      mediaExtras = null
+    } else {
+      mediaExtras = extras
+      setNewMedia(uri)
+    }
+    onPlay()
+  }
+
+  override fun onPlay() {
+    super.onPlay()
+    // TODO 1
+
+    if (ensureAudioFocus()) {
+      mediaSession.isActive = true
+      initializeMediaPlayer()
+      prepareMedia()
+      startPlaying()
+    }
+  }
+
+  override fun onStop() {
+    super.onStop()
+    // TODO 2
+
+    stopPlaying()
+  }
+
+  override fun onPause() {
+    super.onPause()
+    // TODO 3
+
+    pausePlaying()
+  }
+
+  override fun onSeekTo(pos: Long) {
+    super.onSeekTo(pos)
+
+    mediaPlayer?.seekTo(pos.toInt())
+
+    val playbackState: PlaybackStateCompat? = mediaSession.controller.playbackState
+
+    if (playbackState != null) {
+      setState(playbackState.state)
+    } else {
+      setState(PlaybackStateCompat.STATE_PAUSED)
+    }
+  }
+
+  private fun setNewMedia(uri: Uri?) {
+    newMedia = true
+    mediaUri = uri
+  }
+
+  private fun ensureAudioFocus(): Boolean {
+    val audioManager = this.context.getSystemService(
+        Context.AUDIO_SERVICE) as AudioManager
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      val focusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN).run {
+        setAudioAttributes(AudioAttributes.Builder().run {
+          setUsage(AudioAttributes.USAGE_MEDIA)
+          setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
+          build()
+        })
+        build()
+      }
+      this.focusRequest = focusRequest
+      val result = audioManager.requestAudioFocus(focusRequest)
+      return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    } else {
+      val result = audioManager.requestAudioFocus(null,
+          AudioManager.STREAM_MUSIC,
+          AudioManager.AUDIOFOCUS_GAIN)
+      return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  private fun removeAudioFocus() {
+    val audioManager = this.context.getSystemService(
+        Context.AUDIO_SERVICE) as AudioManager
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      focusRequest?.let {
+        audioManager.abandonAudioFocusRequest(it)
+      }
+    } else {
+      audioManager.abandonAudioFocus(null)
+    }
+  }
+
+  private fun initializeMediaPlayer() {
+    // TODO 4
+
+    if (mediaPlayer == null) {
+      mediaPlayer = MediaPlayer()
+      mediaPlayer!!.setOnCompletionListener {
```

Comment:
remove force unwrapping

Reply:
Not a healthy practice, must be updated.

Reply:
This could be a nice place to demonstrate a bug where MediaPlayer returns Null. I was looking for a place to show force return of a method.

Could we leave this in and I'll fix it during Chapter 3.

Reply:
Awesome idea

Reply:
Added to chapter 3 outline.

Reply:
Upon trying to break this down, I realised that this is a bad place to demonstrate force return, the MediaPlayer variable is being set right before the force unwrap and it's an uncompiled class that I couldn't get force return to work in. I'll add a force unwrap somewhere else to demonstrate this debugging technique.